src/openrct2/core/OrcaStream.hpp

```diff
@@ -0,0 +1,685 @@
+    {
+    public:
+        enum class Mode
+        {
+            READING,
+            WRITING,
+        };
+
+        static constexpr uint32_t COMPRESSION_NONE = 0;
+        static constexpr uint32_t COMPRESSION_GZIP = 1;
+
+    private:
+#pragma pack(push, 1)
+        struct Header
+        {
+            uint32_t Magic{};
+            uint32_t TargetVersion{};
+            uint32_t MinVersion{};
+            uint32_t NumChunks{};
+            uint64_t UncompressedSize{};
+            uint32_t Compression{};
+            uint64_t CompressedSize{};
+            std::array<uint8_t, 8> FNV1a{};
+            uint8_t padding[20];
+        };
+        static_assert(sizeof(Header) == 64, "Header should be 64 bytes");
+
+        struct ChunkEntry
+        {
+            uint32_t Id{};
+            uint64_t Offset{};
+            uint64_t Length{};
+        };
+#pragma pack(pop)
+
+        IStream* _stream;
+        Mode _mode;
+        Header _header;
+        std::vector<ChunkEntry> _chunks;
+        MemoryStream _buffer;
+        ChunkEntry _currentChunk;
+
+    public:
+        OrcaStream(IStream& stream, Mode mode)
+        {
+            _stream = &stream;
+            _mode = mode;
+            if (mode == Mode::READING)
+            {
+                _header = _stream->ReadValue<Header>();
+
+                _chunks.clear();
+                for (uint32_t i = 0; i < _header.NumChunks; i++)
+                {
+                    auto entry = _stream->ReadValue<ChunkEntry>();
+                    _chunks.push_back(entry);
+                }
+
+                // Read compressed data into buffer (read in blocks)
+                _buffer = MemoryStream{};
+                uint8_t temp[2048];
+                uint64_t bytesLeft = _header.CompressedSize;
+                do
+                {
+                    auto readLen = std::min(size_t(bytesLeft), sizeof(temp));
+                    _stream->Read(temp, readLen);
+                    _buffer.Write(temp, readLen);
+                    bytesLeft -= readLen;
+                } while (bytesLeft > 0);
+
+                // Uncompress
+                if (_header.Compression == COMPRESSION_GZIP)
+                {
+                    auto uncompressedData = Ungzip(_buffer.GetData(), _buffer.GetLength());
+                    if (_header.UncompressedSize != uncompressedData.size())
+                    {
+                        // Warning?
+                    }
+                    _buffer.Clear();
+                    _buffer.Write(uncompressedData.data(), uncompressedData.size());
+                }
+            }
+            else
+            {
+                _header = {};
+                _header.Compression = COMPRESSION_GZIP;
+
+                _buffer = MemoryStream{};
+            }
+        }
+
+        OrcaStream(const OrcaStream&) = delete;
+
+        ~OrcaStream()
+        {
+            if (_mode == Mode::READING)
+            {
+            }
+            else
```

Comment:
```suggestion
            if (_mode != Mode::READING)
```

Reply:
Or add `throw std::runtime_error("Incorrect mode");` if you want to error.

Reply:
I don't think that a container should typically have such complexity in a destructor other than releasing resources. The write should be made explicit, this container can not be used in parallel properly for this reason.